File: Solutions/2766-find-the-prefix-common-array-of-two-arrays/find-the-prefix-common-array-of-two-arrays.go
func findThePrefixCommonArray(A []int, B []int) []int {
    n := len(A)
    prefixCommonArray := make([]int, n) // Result array to store common counts

    for currentIndex := 0; currentIndex < n; currentIndex++ {
        commonCount := 0

        // Compare elements in A and B within the range of the current prefix
        for aIndex := 0; aIndex <= currentIndex; aIndex++ {
            for bIndex := 0; bIndex <= currentIndex; bIndex++ {
                if A[aIndex] == B[bIndex] {
                    commonCount++         // Increment the count if elements match
                    break                 // Prevent counting duplicates
                }
            }
        }

        // Store the count of common elements for the current prefix
        prefixCommonArray[currentIndex] = commonCount
    }

    return prefixCommonArray
}

File: Solutions/2766-find-the-prefix-common-array-of-two-arrays/find-the-prefix-common-array-of-two-arrays.kt
class Solution {

    fun findThePrefixCommonArray(A: IntArray, B: IntArray): IntArray {

        // Size of arrays
        val n = A.size

        // Answer array
        val result = IntArray(n)

        // Track visited numbers
        // Since numbers are from 1 to n
        val seen = IntArray(n + 1)

        // Current common count
        var common = 0

        // Traverse both arrays
        for (i in 0 until n) {

            // Mark element from A
            seen[A[i]]++

            // If count becomes 2,
            // it means the number appeared
            // in both A and B prefixes
            if (seen[A[i]] == 2) {
                common++
            }

            // Mark element from B
            seen[B[i]]++

            // Again check if it became common
            if (seen[B[i]] == 2) {
                common++
            }

            // Store answer for current prefix
            result[i] = common
        }

        // Return final array
        return result
    }
}